### PR TITLE
Implement conversation turn middleware foundation

### LINKED ISSUE #312: Implement conversation turn middleware foundation
# Problem

LoongClaw's app-layer conversation runtime has strong kernel and registry seams, but cross-cutting turn behavior is still distributed across coordinator, turn loop, and context-engine lifecycle hooks. That shape will become expensive once automation, memory enrichment, prompt shaping, and HITL flows grow further.

# Goal

Introduce a minimal conversation turn middleware foundation in the app layer so new cross-cutting turn behaviors compose through an ordered seam instead of ad hoc branch expansion.

# Scope

- add an ordered turn middleware trait and registry in `crates/app/src/conversation/`
- preserve `ConversationContextEngine` as the authority for prompt assembly
- bridge current context-engine lifecycle hooks through compatibility middleware
- integrate the middleware stack into the existing turn runtime without weakening kernel tool governance
- expose lightweight diagnostics for registered middleware

# Non-Goals

- new channels or UI surfaces
- scheduler / automation productization in this issue
- hosted memory vendors or policy bypasses
- replacing the existing context engine with middleware

# Validation

- focused middleware regression tests fail before implementation and pass after
- `cargo fmt --all -- --check`
- `cargo clippy --workspace --all-targets --all-features -- -D warnings`
- `cargo test --workspace --locked`
- `cargo test --workspace --all-features --locked`

# Context

This issue executes the first implementation slice from the reference comparison and middleware foundation plan prepared on 2026-03-18.

### PR DESCRIPTION
**Pull request by chumyin**
## Summary

- Problem: per-turn conversation lifecycle hooks were hardwired into the runtime flow, so there was no ordered middleware foundation for cross-cutting turn behavior.
- Why it matters: extensible turn processing needs a stable hook model for bootstrap, ingestion, context transformation, after-turn compaction, and delegate child lifecycle events.
- What changed: added ordered conversation turn middleware traits, config-backed registry resolution, runtime selection plumbing, lifecycle hook execution around the major conversation seams, and regression coverage for ordering, runtime snapshots, and delegate lifecycle notifications.
- What did not change (scope boundary): this does not add a marketplace or external plugin transport for middleware, and it does not change provider routing semantics.

## Linked Issues

- Closes #312
- Related # n/a

## Change Type

- [ ] Bug fix
- [x] Feature
- [x] Refactor
- [ ] Documentation
- [ ] Security hardening
- [ ] CI / workflow / release

## Touched Areas

- [ ] Kernel / policy / approvals
- [ ] Contracts / protocol / spec
- [ ] Daemon / CLI / install
- [ ] Providers / routing
- [ ] Tools
- [ ] Browser automation
- [ ] Channels / integrations
- [x] ACP / conversation / session runtime
- [ ] Memory / context assembly
- [x] Config / migration / onboarding
- [ ] Docs / contributor workflow
- [ ] CI / release / workflows

## Risk Track

- [ ] Track A (routine / low-risk)
- [x] Track B (higher-risk / policy-impacting)

If Track B, fill these in:

- Risk notes:
  this inserts a new ordered middleware layer across core conversation lifecycle seams, so ordering and hook boundaries must stay deterministic.
- Rollout / guardrails:
  middleware resolution stays explicit and ordered, and regression coverage exercises ordering, lifecycle execution, and runtime snapshot reporting.
- Rollback path:
  revert the middleware registry and hook wiring to restore the previous hardwired lifecycle flow.

## Validation

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --workspace --all-targets --all-features -- -D warnings`
- [x] `cargo test --workspace --locked`
- [x] `cargo test --workspace --all-features --locked`
- [x] Relevant architecture / dep-graph / docs checks for touched areas
- [x] Additional scenario, benchmark, or manual checks when behavior changed
- [ ] If this changes config/env fallback, limits, or defaults: include before/after behavior and regression coverage for explicit path, fallback path, and boundary values
- [ ] If tests mutate process-global env: document how state is restored or serialized

Commands and evidence:

```text
cargo fmt --all -- --check
- passed

cargo clippy --workspace --all-targets --all-features --locked -- -D warnings
- passed

cargo test --workspace --locked
- passed

cargo test --workspace --all-features --locked
- passed

LOONGCLAW_ARCH_STRICT=true scripts/check_architecture_boundaries.sh
- passed

scripts/check_dep_graph.sh
- passed

scripts/check-docs.sh
- passed

diff CLAUDE.md AGENTS.md
- passed
```

## User-visible / Operator-visible Changes

- conversation runtimes can now resolve ordered turn middleware and run middleware hooks around bootstrap, ingestion, context transformation, after-turn compaction, and delegate child lifecycle events.

## Failure Recovery

- Fast rollback or disable path: revert the middleware foundation to restore the previous lifecycle flow.
- Observable failure symptoms reviewers should watch for: middleware ordering drift, missing lifecycle callbacks, duplicated hook execution, or runtime snapshot reporting diverging from the active middleware set.

## Reviewer Focus

- `crates/app/src/conversation/turn_middleware.rs` and `crates/app/src/conversation/turn_middleware_registry.rs` for the hook contract and ordered resolution behavior.
- `crates/app/src/conversation/runtime.rs` for lifecycle insertion points.
- `crates/app/src/conversation/tests.rs` for ordering, runtime snapshot, and delegate lifecycle regression coverage.

### CODE REVIEW

**Comment by chumyin** [comment]
Follow-up in `854225b` adds the first built-in default turn middleware and tightens registry behavior:

- moved system prompt addition into a built-in `system-prompt-addition` turn middleware instead of keeping it as ad hoc runtime post-processing
- made effective middleware selection include default-enabled built-ins while remaining extensible through config/env overrides
- made the registry reject duplicate middleware ids instead of silently replacing registrations
- added regression coverage for built-in defaults, snapshot reporting, and duplicate-registration failures

Validation on this head:
- `cargo fmt --all -- --check`
- `cargo test -p loongclaw-app --lib --locked -- --test-threads=1`
- `cargo test --workspace --locked`
- `cargo clippy --workspace --all-targets --all-features --locked -- -D warnings`
- `cargo test --workspace --all-features --locked`
- `LOONGCLAW_ARCH_STRICT=true scripts/check_architecture_boundaries.sh`
- `scripts/check_dep_graph.sh`
- `scripts/check-docs.sh`
- `diff CLAUDE.md AGENTS.md`

`cargo deny check advisories bans licenses sources` still reports the existing repository baseline license policy warnings around the `scraper`/`cssparser` chain and was not changed by this work.

**Comment by chumyin** [comment]
Follow-up in `1f512de` moves tool-view system prompt projection into the built-in turn middleware layer instead of leaving it as ad hoc runtime post-processing.

What changed:
- extended the `transform_context` middleware hook with both the runtime default tool view and the requested tool view
- added a second built-in default middleware, `system-prompt-tool-view`, so system prompt projection now lives in the same middleware chain as system prompt additions
- replaced the duplicated runtime rewrite path with a shared `build_context_for_tool_view(...)` flow
- expanded regression coverage for built-in middleware defaults, runtime metadata, snapshot selection, and restricted tool-view prompt projection

Validation on this head:
- `cargo fmt --all -- --check`
- `cargo test -p loongclaw-app --lib --locked -- --test-threads=1`
- `cargo clippy --workspace --all-targets --all-features --locked -- -D warnings`
- `cargo test --workspace --locked`
- `cargo test --workspace --all-features --locked`
- `LOONGCLAW_ARCH_STRICT=true scripts/check_architecture_boundaries.sh`
- `scripts/check_dep_graph.sh`
- `scripts/check-docs.sh`
- `diff CLAUDE.md AGENTS.md`

`cargo deny check advisories bans licenses sources` still reports the existing repository baseline MPL-2.0/license-policy failures in the `scraper` / `cssparser` chain and unchanged duplicate-lock warnings; this diff did not touch dependency manifests or lockfiles.

**Comment by chumyin** [comment]
Rebased this branch onto `dev` and force-pushed the updated history. The PR head is now `9432024cb5198942eaf1aec0cd7ccd7194b9aeed`.

Local validation rerun on the rebased head:
- `cargo fmt --all -- --check`
- `cargo test -p loongclaw-app --lib --locked -- --test-threads=1`
- `cargo clippy --workspace --all-targets --all-features --locked -- -D warnings`
- `cargo test --workspace --locked`
- `cargo test --workspace --all-features --locked`
- `LOONGCLAW_ARCH_STRICT=true scripts/check_architecture_boundaries.sh`
- `scripts/check_dep_graph.sh`
- `scripts/check-docs.sh`
- `diff CLAUDE.md AGENTS.md`
- `cargo deny check advisories bans licenses sources`

`cargo deny` still reports the existing baseline warnings only (the unmatched `Unicode-DFS-2016` allowance and the known duplicate crate entries in `Cargo.lock`); no new deny failures were introduced by this branch.

File: crates/app/src/conversation/tests.rs [hunk 1]
```diff
@@ -11455,6 +11724,105 @@ async fn handle_turn_with_runtime_executes_delegate_via_coordinator() {
...
+                child_session_id.clone(),
+            ),
+            (
+                "on_subagent_ended".to_owned(),
+                "root-session".to_owned(),
+                child_session_id,
+            ),
+        ]
+    );
+}
```
[the remark this answers is not in the document]
Comment (chumyin) [reply crates/app/src/conversation/tests.rs:13852]: Checked this one and left it as-is. The test already uses a unique per-run sqlite filename and proactively removes any preexisting file before it starts, so it does not create cross-test interference. I did not add a tail cleanup-only line because that still would not be panic-safe, and adding a dedicated temp-file guard just for this case would be extra churn for a non-failing minor issue.

File: crates/app/src/conversation/runtime.rs [hunk 2]
```diff
@@ -282,48 +355,225 @@ pub fn collect_context_engine_runtime_snapshot(
...
     }
+
+    pub fn with_turn_middlewares(
+        turn_middlewares: Vec<Box<dyn ConversationTurnMiddleware>>,
+    ) -> Self {
+        Self {
+            context_engine: DefaultContextEngine,
+            turn_middlewares,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (chumyin) [reply crates/app/src/conversation/runtime.rs:377]: Checked this against the current call sites and left it unchanged. These constructors are only used for explicit in-crate runtime injection, and the replacement semantics are intentional because the tests need to control the complete middleware chain deterministically. The additive path remains the config/env resolution path (`resolve_turn_middleware_selection` plus `from_config_or_env`), so appending defaults here would silently change the test-only builder behavior rather than fix a production bug.

File: crates/app/src/conversation/turn_middleware_registry.rs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    if let Ok(mut guard) = env_override().lock() {
+        *guard = Some(value.map(str::to_owned));
+    }
+}
+
+#[cfg(test)]
+pub(crate) fn clear_turn_middleware_env_override() {
+    if let Ok(mut guard) = env_override().lock() {
+        *guard = None;
+    }
```
[the remark this answers is not in the document]
Comment (chumyin) [reply crates/app/src/conversation/turn_middleware_registry.rs:228]: Fixed the env-isolation part in `3b741e8`: the turn-middleware override tests now use a shared process-wide selector lock plus RAII restoration, so the override no longer leaks on panic or interleaves with the context-engine selector tests. I did not add a registry reset or unregister API because the registry tests already use distinct middleware ids, so there is no demonstrated duplicate-id collision across tests, and adding more mutable test-only registry surface would be broader than the failing case here.

**Comment by chumyin** [comment]
Follow-up fix landed in `3b741e8`.

What changed:
- moved turn-middleware factory execution out of the registry read-lock path
- unified the process-wide selector env test lock used by context-engine and turn-middleware tests
- switched the remaining turn-middleware env test path to RAII restoration and added regression coverage for both lock scope and env-lock sharing

Local validation on `3b741e8`:
- `cargo fmt --all -- --check`
- `cargo test -p loongclaw-app --lib --locked -- --test-threads=1`
- `cargo clippy --workspace --all-targets --all-features --locked -- -D warnings`
- `cargo test --workspace --locked`
- `cargo test --workspace --all-features --locked`
- `LOONGCLAW_ARCH_STRICT=true scripts/check_architecture_boundaries.sh`
- `scripts/check_dep_graph.sh`
- `scripts/check-docs.sh`
- `diff CLAUDE.md AGENTS.md`
- `cargo deny check advisories bans licenses sources`

GitHub CI is green on this head as well. The remaining PR gate is review approval.

**Comment by chumyin** [comment]
rebased onto current dev and tightened the middleware foundation before merge: preserved the built-in defaults when extending the runtime chain, centralized the built-in middleware spec, kept the delegate runtime contract prompt wiring intact through the rebase, and made duplicate registrations fail before invoking replacement factories. rechecked with fmt, clippy, focused middleware regressions, workspace tests, and the repo architecture, dep graph, and docs checks.
